docs/docs/en/configuration.md
<!--
  ~ Copyright (C) 2025 Enedis Smarties team <dt-dsi-nexus-lab-smarties@enedis.fr>
  ~ 
  ~ SPDX-FileContributor: Jehan BOUSCH
  ~ 
  ~ SPDX-License-Identifier: Apache-2.0
-->

# Configuration

## Introduction

The **tic4eebus** application configuration is organized into 7 sections:

|Name|Description|Format|Value|
|---|-----------|------|------|
| *OverloadProtection* | Overload protection parameters | Structure | [Overload protection](#overload-protection) values |
| *Vehicle* | Electric vehicle parameters | Structure | [Electric vehicle](#electric-vehicle) values |
| *Wallbox* | Charging station parameters | Structure | [Charging station](#charging-station) values |
| *Log* | Logging parameters | Structure | [Logging parameters](#logging) values |
| *DataModel* | Data model parameters | Structure | [Data model](#data-model) values |
| *TeleInformationClient* | Tele-information client (TIC) parameters | Structure | [TIC](#tele-information-client-tic) values |
| *Eebus* | EEBUS protocol parameters | Structure | [EEBUS protocol](#eebus-protocol) values |

Each section above is stored in the configuration file in YAML format.

The configuration file is loaded at application startup with the default path "*examples/config.yaml*".

This configuration file path can be modified via the application command line with the "*configFilePath*" argument.

## Description

### Overload Protection

The overload protection configuration includes the following parameters:

|Name|Description|Format|Value|
|---|-----------|------|------|
| *Enable* | Algorithm activation | Boolean | - **true** if the algorithm is active <br>- **false** otherwise |
| *RunningPeriodInSeconds* | Algorithm execution period in seconds | Integer | From 0 to 2^31 - 1 |
| *CurrentLimit* | Current limitation parameters | Structure | [Current limitation](#current-limitation) values |

#### Current Limitation

The current limitation parameters configuration for overload protection includes the following fields:

|Name|Description|Format|Value|
|---|-----------|------|------|
| *ValueInAmps* | EV charging current limit in Amperes | Decimal | From 0.0 to 1.79 × 10^308 |
| *LockDelayInSeconds* | Limitation lock duration in seconds | Decimal | From 0 to 1.79 × 10^308 |

<u>Notes:</u>

- The EV charging current limit (*ValueInAmps*) is only used if the overload protection algorithm is inactive (*Enable*=**false**)
- The limitation lock duration (*LockDelayInSeconds*) is only used if the overload protection algorithm is active (*Enable*=**true**)

### Electric Vehicle

The electric vehicle parameters configuration includes the following fields:

|Name|Description|Format|Value|
|---|-----------|------|------|
| *UpdateDataPeriodInSeconds* | EV data update period in seconds | Integer | From 0 to 2^31 - 1 <br> 0 means no periodic update |
| *DataPersistent* | EV data persistence indicator | Boolean | - **true** if data is persistent <br>- **false** otherwise |

### Charging Station

The charging station parameters configuration includes the following fields:

|Name|Description|Format|Value|
|---|-----------|------|------|
| *UpdateDataPeriodInSeconds* | Station data update period in seconds | Integer | From 0 to 2^31 - 1 <br> 0 means no periodic update |
| *DataPersistent* | Station data persistence indicator | Boolean | - **true** if data is persistent <br>- **false** otherwise |

### Logging

The logging parameters configuration includes the following fields:

|Name|Description|Format|Value|
|---|-----------|------|------|
| *Level* | Logging level used | Enumeration | [Logging level](#logging-level) values |
| *FilePath* | Logging file path | String | Example: "var/log/tic4eebus.log" |
| *Rotation* | Logging file rotation parameters | Structure | [File rotation](#file-rotation) values |

### Data Model

The data model parameters configuration includes the following fields:

|Name|Description|Format|Value|
|---|-----------|------|------|
| *Csv* | CSV file parameters | Structure | [CSV file](#csv-file) values |
| *InfluxDb* | InfluxDB database parameters | Structure | [InfluxDB database](#influxdb-database) values |

<u>Notes:</u>

The *Csv* and *InfluxDb* fields are optional.
If the *Csv* field does not exist, there is no CSV file.
If the *InfluxDb* field does not exist, the InfluxDB database is not populated.

#### CSV File

The CSV file parameters configuration includes the following fields:

|Name|Description|Format|Value|
|---|-----------|------|------|
| *FilePath* | CSV file path | String | Example: "var/data/DataModel.csv" |
| *Rotation* | CSV file rotation parameters | Structure | [File rotation](#file-rotation) values |

#### InfluxDB Database

The InfluxDB database parameters configuration includes the following fields:

|Name|Description|Format|Value|
|---|-----------|------|------|
| *Bucket* | InfluxDB database name (called bucket) | String | Example: "demo-bucket" |
| *Org* | Organization name associated with the database and tables (called measurements) | String | Example: "demo-org" |
| *Token* | Authentication key used for database connection | String | Example: "He_nwgoqXu4XggIzaiUWFHALKnS5JskdTLzGlSYeJMNkjCD-pyR6Yc5Hvl8NWj5Qo5C80mLvuJAS1IuqOcq4GQZZ" |
| *IpAddress* | IP address of the server hosting the InfluxDB database | String | Example: "127.0.0.1" |
| *TcpPort* | TCP port of the server hosting the InfluxDB database | Integer | From 1 to 65535 |

### Tele-Information Client (TIC)

The tele-information client parameters configuration includes the following fields:

|Name|Description|Format|Value|
|---|-----------|------|------|
| *Tic2Websocket* | TIC2WebSocket client parameters | Structure | [Tic2WebSocket](#tic2websocket) values |
| *TicIdentifier* | Linky meter identification parameters | Structure | [TicIdentifier](#ticidentifier) values |

#### Tic2Websocket

The Tic2Websocket client parameters configuration includes the following fields:

|Name|Description|Format|Value|
|---|-----------|------|------|
| *IpAddress* | IP address of the server hosting the Tic2Websocket application | String | Example: "127.0.0.1" |
| *TcpPort* | TCP port of the server hosting the Tic2Websocket application | Integer | From 1 to 65535 |

#### TicIdentifier

The Linky meter identification parameters configuration includes the following fields:

|Name|Description|Format|Value|
|---|-----------|------|------|
| *SerialNumber* | Linky meter serial number | 12-character string | Example: "041976216986" |

<u>Note:</u>

If the *SerialNumber* field is empty, the application will use the first serial number provided by the *Tic2Websocket* application.

### EEBUS Protocol

The EEBUS protocol parameters configuration includes the following fields:

|Name|Description|Format|Value|
|---|-----------|------|------|
| *ServerPort* | TCP port of the charging station EEBUS server | Integer | From 1 to 65535 |
| *RemoteSki* | Identification key of the charging station EEBUS server | 40 hexadecimal character string | Example: "50abfe7714d034b8b15e488b91831047657b9ff2" |
| *CertificateFilePath* | Security certificate path for the energy management system | String | Example: "examples/energy-guard.cert" |
| *PrivateKeyFilePath* | Energy management system encryption private key | String | Example: "examples/energy-guard.key" |
| *VendorCode* | Energy management company identification code compliant with [IANA PEN](https://www.iana.org/assignments/enterprise-numbers/) | String | Example: "i:54076" |
| *DeviceBrand* | Energy management system brand name | String | Example: "Enedis" |
| *DeviceModel* | Energy management system model name | String | Example: "PAC" |
| *SerialNumber* | Energy management system serial number | String | Example: "12345678" |
| *HeartbeatTimeoutInSeconds* | Maximum delay allowed for sending [EEBUS heartbeat](../var/references/EEBus_UC_TS_OverloadProtectionByEVChargingCurrentCurtailment_V1.0.1b.pdf#page=34) to the charging station | Integer | From 0 to 2^32 -1 |

<u>Note:</u>

If the *HeartbeatTimeoutInSeconds* parameter is configured with a value greater than 4 seconds, the charging station, in accordance with [OPEV use case scenario 2](../var/references/EEBus_UC_TS_OverloadProtectionByEVChargingCurrentCurtailment_V1.0.1b.pdf#page=45), considers that the energy management system is not available and sets the charge limitation to its minimum value.

## Types

### Logging Level

The logging level defines the minimum threshold for information to be kept in logs.

It can take the following values, sorted in descending order of severity:

|Enumeration Value|Enumeration Description|
|-----------------------|----------------------------|
| *panic* | Highest level of severity. Logs and then call [panic()](https://go.dev/blog/defer-panic-and-recover). |
| *fatal* | Critical severity level. Logs and then exits the program. |
| *error* | Level used for errors that should definitely be noted. |
| *warn* | Level for non-critical information that deserve eyes |
| *info* | Level for general information about what's going on inside the application. |
| *debug* | Level only enabled when debugging. Very verbose logging. |
| *trace* | Level designating finer-grained informational events than the *debug* |

### File Rotation

File rotation includes the following fields:

|Name|Description|Format|Value|
|---|-----------|------|------|
| *PeriodInHours* | Duration of a rotation period in hours | Integer | From 0 to 2^31 - 1 |
| *PeriodCount* | Maximum number of rotations | Integer | From 0 to 2^31 - 1 |
| *PeriodPattern* | Pattern used for the file name at each new rotation | String | |

## Example

Here is a complete example of the default [configuration file](../var/examples/config.yaml) in YAML format:

```yaml
# Overload protection configuration
OverloadProtection:
  Enable : true
  RunningPeriodInSeconds : 1
  CurrentLimit :
    ValueInAmps : 16.0 # Used only if Enable is false
    LockDelayInSeconds : 10.0 # Used only if Enable is true

# Vehicle configuration
Vehicle:
  UpdateDataPeriodInSeconds : 5 # Unused if 0
  DataPersistent: false

# Wallbox configuration
Wallbox:
  UpdateDataPeriodInSeconds : 5 # Unused if 0
  DataPersistent: false

# Log configuration
Log:
  Level : "trace" # Available levels : panic,fatal,error,warn,info,debug,trace
  FilePath : "var/log/tic4eebus.log"
  Rotation:
    PeriodInHours : 24 # Each day a new file is created
    PeriodCount : 15 # 15 files maximum are created (15 days of logs)

CHANGELOG.md
<!--
  ~ Copyright (C) 2025 Enedis Smarties team <dt-dsi-nexus-lab-smarties@enedis.fr>
  ~ 
  ~ SPDX-FileContributor: Jehan BOUSCH
  ~ SPDX-FileContributor: Mathieu SABARTHES
  ~ 
  ~ SPDX-License-Identifier: Apache-2.0
-->
# Journal des modifications

[🇫🇷 Français](CHANGELOG.fr.md) | [🇺🇸 English](CHANGELOG.md)

## [v1.1.0](https://github.com/Enedis-OSS/tic4eebus/tree/v1.1.0)
### ✨ Nouvelles fonctionnalités:
* Écriture du modèle de données vers InfluxDB pour la visualisation avec Grafana
### 🔧 Améliorations techniques:
* Configuration optionnelle pour le modèle de données (fichier CSV, InfluxDB)

## [v1.0.0](https://github.com/Enedis-OSS/tic4eebus/tree/v1.0.0)
### ✨ Nouvelles fonctionnalités :
* Cas d'utilisation OPEV EEBUS (scénario 1) : Le gestionnaire d'énergie limite le courant de charge du VE
* Cas d'utilisation OPEV EEBUS (scénario 2) : Le VE vérifie la disponibilité du gestionnaire d'énergie
* Cas d'utilisation OPEV EEBUS (scénario 3) : Le gestionnaire d'énergie envoie un état d'erreur
* Fichier de configuration avec paramètres pour la protection contre les surcharges, VE, logs, modèle de données, télé-information client compteur Linky, EEBUS
* Fichier de log avec rotation
* Interpréteur de ligne de commande avec aide, version et chemin du fichier de configuration
* Modèle de données exporté en fichier CSV avec rotation

config/config.go
// SPDX-FileCopyrightText: 2025 Enedis Smarties team <dt-dsi-nexus-lab-smarties@enedis.fr>
// SPDX-FileContributor: Jehan BOUSCH
//
// SPDX-License-Identifier: Apache-2.0

package config

import (
	"fmt"
	"os"
	"reflect"
	"regexp"

	log "github.com/sirupsen/logrus"
	"gopkg.in/yaml.v2"
)

type CurrentLimitConfig struct {
	ValueInAmps        float64
	LockDelayInSeconds float64
}

type FileRotationConfig struct {
	PeriodInHours int
	PeriodCount   int
	PeriodPattern string
}

type CsvConfig struct {
	FilePath string
	Rotation FileRotationConfig
}

type InfluxDbConfig struct {
	Bucket    string
	Org       string
	Token     string
	IpAddress string
	TcpPort   int
}

type TIC2WebsocketConfig struct {
	IPAddress string
	TCPPort   int
}

type TICIdentifierConfig struct {
	SerialNumber string
}

type OverloadProtectionConfig struct {
	Enable                 bool
	RunningPeriodInSeconds int
	CurrentLimit           CurrentLimitConfig
}

type VehicleConfig struct {
	UpdateDataPeriodInSeconds int
	DataPersistent            bool
}

type WallboxConfig struct {
	UpdateDataPeriodInSeconds int
	DataPersistent            bool
}

type LogConfig struct {
	Level    log.Level
	FilePath string
	Rotation FileRotationConfig
}

type DataModelConfig struct {
	Csv      *CsvConfig
	InfluxDb *InfluxDbConfig
}

type TeleInformationClientConfig struct {
	TIC2Websocket TIC2WebsocketConfig
	TICIdentifier TICIdentifierConfig
}

type EEBUSConfig struct {
	ServerPort                int
	RemoteSKI                 string
	CertificateFilePath       string
	PrivateKeyFilePath        string
	VendorCode                string
	DeviceBrand               string
	DeviceModel               string
	SerialNumber              string
	HeartbeatTimeoutInSeconds int
}

type Config struct {
	OverloadProtection    OverloadProtectionConfig
	Vehicle               VehicleConfig
	Wallbox               WallboxConfig
	Log                   LogConfig
	DataModel             DataModelConfig
	TeleInformationClient TeleInformationClientConfig
	EEBUS                 EEBUSConfig
}

const (
	INVALID_PARAMETER = "invalid config parameter"
	REMOTE_SKI_REGEXP = "^[a-fA-F0-9]{40}$" // SHA-1 hash is 20 bytes, represented as 40 hexadecimal digits
)

func LoadConfig(configFilePath string) (Config, error) {
	var config Config
	configMap := make(map[any]any)
	data, err := os.ReadFile(configFilePath)
	if err != nil {
		return config, fmt.Errorf("error reading YAML file: %v", err)
	}
	dataWithEnv := os.ExpandEnv(string(data))
	err = yaml.Unmarshal([]byte(dataWithEnv), &configMap)
	if err != nil {
		return config, fmt.Errorf("error unmarshalling YAML: %v", err)
	}
	config.OverloadProtection.Enable,
		config.OverloadProtection.RunningPeriodInSeconds,
		config.OverloadProtection.CurrentLimit.ValueInAmps,
		config.OverloadProtection.CurrentLimit.LockDelayInSeconds,
		err = loadOverloadProtection(configMap)
	if err != nil {
		return config, err
	}
	config.Vehicle.UpdateDataPeriodInSeconds,
		config.Vehicle.DataPersistent,
		err = loadVehicle(configMap)
	if err != nil {
		return config, err
	}
	config.Wallbox.UpdateDataPeriodInSeconds,
		config.Wallbox.DataPersistent,
		err = loadWallbox(configMap)
	if err != nil {
		return config, err
	}
	config.Log.Level,
		config.Log.FilePath,
		config.Log.Rotation.PeriodInHours,
		config.Log.Rotation.PeriodCount,
		config.Log.Rotation.PeriodPattern,
		err = loadLog(configMap)
	if err != nil {
		return config, err
	}
	config.DataModel.Csv,
		config.DataModel.InfluxDb,
		err = loadDataModel(configMap)
	if err != nil {
		return config, err
	}
	config.TeleInformationClient.TIC2Websocket.IPAddress,
		config.TeleInformationClient.TIC2Websocket.TCPPort,
		config.TeleInformationClient.TICIdentifier.SerialNumber,
		err = loadTeleInformationClient(configMap)
	if err != nil {
		return config, err
	}
	config.EEBUS.ServerPort,
		config.EEBUS.RemoteSKI,
		config.EEBUS.CertificateFilePath,
		config.EEBUS.PrivateKeyFilePath,
		config.EEBUS.VendorCode,
		config.EEBUS.DeviceBrand,
		config.EEBUS.DeviceModel,
		config.EEBUS.SerialNumber,
		config.EEBUS.HeartbeatTimeoutInSeconds,
		err = loadEEBUS(configMap)
	if err != nil {
		return config, err
	}

	return config, err
}

func DumpConfig(config Config) (string, error) {
	buffer, err := yaml.Marshal((&config))

	if err != nil {
		return "", err
	}

	return string(buffer), err
}

func loadOverloadProtection(configMap map[interface{}]interface{}) (enable bool, runningPeriodInSeconds int, valueInAmps float64, lockDelayInSeconds float64, err error) {
	paramParentName, paramName := "", "OverloadProtection"
	overloadProtectionMap, err := loadParameterAsMap(configMap, paramParentName, paramName, true)
	if err != nil {
		return false, 0, 0.0, 0.0, err
	}
	paramParentName = paramName
	enable, err = loadParameterAsBoolean(overloadProtectionMap, paramParentName, "Enable")
	if err != nil {
		return enable, 0, 0.0, 0.0, err
	}
	runningPeriodInSeconds, err = loadParameterAsInt(overloadProtectionMap, paramParentName, "RunningPeriodInSeconds", true, 0, false, 0)
	if err != nil {
		return enable, runningPeriodInSeconds, 0.0, 0.0, err
	}
	paramName = "CurrentLimit"
	currentLimitMap, err := loadParameterAsMap(overloadProtectionMap, paramParentName, paramName, true)
	if err != nil {
		return enable, runningPeriodInSeconds, 0.0, 0.0, err
	}
	paramParentName = paramParentName + "." + paramName
	valueInAmps, err = loadParameterAsFloat(currentLimitMap, paramParentName, "ValueInAmps", true, 0.0, false, 0.0)
	if err != nil {
		return enable, runningPeriodInSeconds, 0.0, 0.0, err
	}
	lockDelayInSeconds, err = loadParameterAsFloat(currentLimitMap, paramParentName, "LockDelayInSeconds", true, 0.0, false, 0.0)

	return enable, runningPeriodInSeconds, valueInAmps, lockDelayInSeconds, err
}

func loadVehicle(configMap map[interface{}]interface{}) (updateDataPeriodInSeconds int, dataPersistent bool, err error) {
	paramParentName, paramName := "", "Vehicle"
	vehicleMap, err := loadParameterAsMap(configMap, paramParentName, paramName, true)
	if err != nil {
		return 0, false, err
	}
	paramParentName = paramName
	updateDataPeriodInSeconds, err = loadParameterAsInt(vehicleMap, paramParentName, "UpdateDataPeriodInSeconds", true, 0, false, 0)
	if err != nil {
		return 0, false, err
	}
	dataPersistent, err = loadParameterAsBoolean(vehicleMap, paramParentName, "DataPersistent")

	return updateDataPeriodInSeconds, dataPersistent, err
}

func loadWallbox(configMap map[interface{}]interface{}) (updateDataPeriodInSeconds int, dataPersistent bool, err error) {
	paramParentName, paramName := "", "Wallbox"
	wallboxMap, err := loadParameterAsMap(configMap, paramParentName, paramName, true)
	if err != nil {
		return 0, false, err
	}
	paramParentName = paramName
	updateDataPeriodInSeconds, err = loadParameterAsInt(wallboxMap, paramParentName, "UpdateDataPeriodInSeconds", true, 0, false, 0)
	if err != nil {
		return 0, false, err
	}
	dataPersistent, err = loadParameterAsBoolean(wallboxMap, paramParentName, "DataPersistent")

	return updateDataPeriodInSeconds, dataPersistent, err
}

func loadLog(configMap map[interface{}]interface{}) (level log.Level, filePath string, periodInHours int, periodCount int, periodPattern string, err error) {
	paramParentName, paramName := "", "Log"
	logMap, err := loadParameterAsMap(configMap, paramParentName, paramName, true)
	if err != nil {
		return 0, "", 0, 0, "", err
	}
	paramParentName = paramName
	level, err = loadParameterAsLogLevel(logMap, paramParentName, "Level")
	if err != nil {
		return 0, "", 0, 0, "", err
	}
	filePath, err = loadParameterAsString(logMap, paramParentName, "FilePath", true)
	if err != nil {
		return level, "", 0, 0, "", err
	}
	periodInHours, periodCount, periodPattern, err = loadParameterAsFileRotationMap(logMap, paramParentName, "Rotation")
	if err != nil {
		return level, filePath, 0, 0, "", err
	}

	return level, filePath, periodInHours, periodCount, periodPattern, err
}

func loadDataModel(configMap map[interface{}]interface{}) (csvConfig *CsvConfig, influxDbConfig *InfluxDbConfig, err error) {
	paramParentName, paramName := "", "DataModel"
	dataModelMap, err := loadParameterAsMap(configMap, "", "DataModel", true)
	if err != nil {
		return nil, nil, err
	}
	paramParentName = paramName
	paramName = "Csv"
	csvMap, err := loadParameterAsMap(dataModelMap, paramParentName, paramName, false)
	if err != nil {
		return nil, nil, err
	}
	if csvMap != nil {
		paramParentName = paramParentName + "." + paramName
		filePath, err := loadParameterAsString(csvMap, paramParentName, "FilePath", true)
		if err != nil {
			return nil, nil, err
		}
		periodInHours, periodCount, periodPattern, err := loadParameterAsFileRotationMap(csvMap, paramParentName, "Rotation")
		if err != nil {
			return nil, nil, err
		}
		csvConfig = &CsvConfig{filePath, FileRotationConfig{periodInHours, periodCount, periodPattern}}
	}
	paramParentName = "DataModel"
	paramName = "InfluxDb"
	influxDbMap, err := loadParameterAsMap(dataModelMap, paramParentName, paramName, false)
	if err != nil {
		return nil, nil, err
	}
	if len(influxDbMap) > 0 {
		paramParentName = paramParentName + "." + paramName
		bucket, err := loadParameterAsString(influxDbMap, paramParentName, "Bucket", true)
		if err != nil {
			return nil, nil, err
		}
		org, err := loadParameterAsString(influxDbMap, paramParentName, "Org", true)
		if err != nil {
			return nil, nil, err
		}
		token, err := loadParameterAsString(influxDbMap, paramParentName, "Token", true)
		if err != nil {
			return nil, nil, err
		}
		ipAddress, err := loadParameterAsString(influxDbMap, paramParentName, "IpAddress", true)
		if err != nil {
			return nil, nil, err
		}
		tcpPort, err := loadParameterAsInt(influxDbMap, paramParentName, "TcpPort", true, 1, true, 65535)
		if err != nil {
			return nil, nil, err
		}
		influxDbConfig = &InfluxDbConfig{bucket, org, token, ipAddress, tcpPort}
	}

	return csvConfig, influxDbConfig, err
}

func loadTeleInformationClient(configMap map[interface{}]interface{}) (tic2WebsocketIPAddress string, tic2WebsocketTCPPort int, ticIdentifierSerialNumber string, err error) {
	paramParentName, paramName := "", "TeleInformationClient"
	teleInformationClientMap, err := loadParameterAsMap(configMap, paramParentName, paramName, true)
	if err != nil {
		return "", 0, "", err
	}
	paramParentName = paramName
	tic2WebsocketMap, err := loadParameterAsMap(teleInformationClientMap, paramParentName, "TIC2Websocket", true)
	if err != nil {
		return "", 0, "", err
	}
	paramParentName = paramParentName + "." + paramName
	tic2WebsocketIPAddress, err = loadParameterAsString(tic2WebsocketMap, paramParentName, "IPAddress", true)
	if err != nil {
		return "", 0, "", err
	}
	tic2WebsocketTCPPort, err = loadParameterAsInt(tic2WebsocketMap, paramParentName, "TCPPort", true, 1, true, 65535)
	if err != nil {
		return tic2WebsocketIPAddress, 0, "", err
	}
	ticIdentifierMap, err := loadParameterAsMap(teleInformationClientMap, paramName, "TICIdentifier", true)
	if err != nil {
		return tic2WebsocketIPAddress, tic2WebsocketTCPPort, "", err
	}
	ticIdentifierSerialNumber, err = loadParameterAsString(ticIdentifierMap, paramParentName, "SerialNumber", false)

	return tic2WebsocketIPAddress, tic2WebsocketTCPPort, ticIdentifierSerialNumber, err
}

func loadEEBUS(configMap map[interface{}]interface{}) (serverPort int,
	remoteSKI string,
	certificateFilePath string,
	privateKeyFilePath string,
	vendorCode string,
	deviceBrand string,
	deviceModel string,
	serialNumber string,
	heartbeatTimeoutInSeconds int,
	err error) {

	paramParentName, paramName := "", "EEBUS"
	eebusMap, err := loadParameterAsMap(configMap, paramParentName, paramName, true)
	if err != nil {
		return 0, "", "", "", "", "", "", "", 0, err
	}
	paramParentName = paramName
	serverPort, err = loadParameterAsInt(eebusMap, paramParentName, "ServerPort", true, 1, true, 65535)
	if err != nil {
		return 0, "", "", "", "", "", "", "", 0, err
	}
	remoteSKI, err = loadParameterAsSKI(eebusMap, paramParentName, "RemoteSKI")
	if err != nil {
		return 0, "", "", "", "", "", "", "", 0, err
	}
	certificateFilePath, err = loadParameterAsExistingFilePath(eebusMap, paramParentName, "CertificateFilePath")
	if err != nil {
		return 0, "", "", "", "", "", "", "", 0, err
	}
	privateKeyFilePath, err = loadParameterAsExistingFilePath(eebusMap, paramParentName, "PrivateKeyFilePath")
	if err != nil {
		return 0, "", "", "", "", "", "", "", 0, err
	}
	vendorCode, err = loadParameterAsString(eebusMap, paramParentName, "VendorCode", true)
	if err != nil {
		return 0, "", "", "", "", "", "", "", 0, err
	}
	deviceBrand, err = loadParameterAsString(eebusMap, paramParentName, "DeviceBrand", true)
	if err != nil {
		return 0, "", "", "", "", "", "", "", 0, err
	}
	deviceModel, err = loadParameterAsString(eebusMap, paramParentName, "DeviceModel", true)
	if err != nil {
		return 0, "", "", "", "", "", "", "", 0, err
	}
	serialNumber, err = loadParameterAsString(eebusMap, paramParentName, "SerialNumber", true)
	if err != nil {
		return 0, "", "", "", "", "", "", "", 0, err
	}
	heartbeatTimeoutInSeconds, err = loadParameterAsInt(eebusMap, paramParentName, "HeartbeatTimeoutInSeconds", true, 0, false, 0)
	if err != nil {
		return 0, "", "", "", "", "", "", "", 0, err
	}

	return serverPort, remoteSKI, certificateFilePath, privateKeyFilePath, vendorCode, deviceBrand, deviceModel, serialNumber, heartbeatTimeoutInSeconds, nil
}

func loadParameterAsMap(parameterMap map[interface{}]interface{}, paramParentName string, paramName string, mandatory bool) (map[interface{}]interface{}, error) {
	parameter, ok := parameterMap[paramName]
	if !ok {
		if mandatory {
			if len(paramParentName) == 0 {
				return nil, fmt.Errorf("%s: %s not found", INVALID_PARAMETER, paramName)
			} else {
				return nil, fmt.Errorf("%s: %s.%s not found", INVALID_PARAMETER, paramParentName, paramName)
			}
		} else {
			return nil, nil
		}
	}
	parameterValue, ok := parameter.(map[any]any)
	if parameterValue == nil {
		return nil, nil
	}

	if !ok {
		return nil, fmt.Errorf("%s: %s.%s is not a map (%v)", INVALID_PARAMETER, paramParentName, paramName, reflect.TypeOf(parameter))
	}

	return parameterValue, nil
}

func loadParameterAsExistingFilePath(parameterMap map[interface{}]interface{}, paramParentName string, paramName string) (string, error) {
	filePath, err := loadParameterAsString(parameterMap, paramParentName, paramName, true)
	if err != nil {
		return "", err
	}
	_, err = os.Stat(filePath)
	if err != nil {
		if os.IsNotExist(err) {
			return "", fmt.Errorf("%s: %s.%s file '%s' not found", INVALID_PARAMETER, paramParentName, paramName, filePath)
		} else {
			return "", fmt.Errorf("%s: %s.%s file '%s' error (%s)", INVALID_PARAMETER, paramParentName, paramName, filePath, err.Error())
		}
	}

	return filePath, nil
}

func loadParameterAsFileRotationMap(parentMap map[interface{}]interface{}, paramParentName string, paramName string) (periodInHours int, periodCount int, periodPattern string, err error) {
	rotationMap, err := loadParameterAsMap(parentMap, paramParentName, paramName, true)
	if err != nil {
		return 0, 0, "", err
	}
	newParentKey := paramParentName + "." + paramName
	periodInHours, err = loadParameterAsInt(rotationMap, newParentKey, "PeriodInHours", true, 0, false, 0)
	if err != nil {
		return 0, 0, "", err
	}
	periodCount, err = loadParameterAsInt(rotationMap, newParentKey, "PeriodCount", true, 0, false, 0)
	if err != nil {
		return 0, 0, "", err
	}
	periodPattern, err = loadParameterAsString(rotationMap, newParentKey, "PeriodPattern", true)
	if err != nil {
		return 0, 0, "", err
	}

	return periodInHours, periodCount, periodPattern, nil
}

func loadParameterAsLogLevel(parameterMap map[interface{}]interface{}, paramParentName string, paramName string) (log.Level, error) {
	level, err := loadParameterAsString(parameterMap, paramParentName, paramName, true)
	if err != nil {
		return 0, err
	}
	logLevel, err := log.ParseLevel(level)
	if err != nil {
		return 0, fmt.Errorf("%s: %s.%s unknown (%s)", INVALID_PARAMETER, paramParentName, paramName, err.Error())
	}

	return logLevel, nil
}

func loadParameterAsSKI(parameterMap map[interface{}]interface{}, paramParentName string, paramName string) (string, error) {
	ski, err := loadParameterAsString(parameterMap, paramParentName, paramName, true)
	if err != nil {
		return "", err
	}
	re, err := regexp.Compile(REMOTE_SKI_REGEXP)
	if err != nil {
		return "", fmt.Errorf("%s: %s.%s regex error (%s)", INVALID_PARAMETER, paramParentName, paramName, err.Error())
	}
	if len(ski) != 40 { // SHA-1 hash is 20 bytes, represented as 40 hexadecimal digits
		return "", fmt.Errorf("%s: %s.%s must be 40 hexadecimal digits", INVALID_PARAMETER, paramParentName, paramName)
	}
	// Check if the string matches the regex for a SHA-1 hash
	if !re.MatchString(ski) {
		return "", fmt.Errorf("%s: %s.%s is not a SHA-1 hash with 20 hexadecimal digits", INVALID_PARAMETER, paramParentName, paramName)
	}

	return ski, nil
}

func loadParameterAsBoolean(parameterMap map[interface{}]interface{}, paramParentName string, paramName string) (bool, error) {
	parameter, ok := parameterMap[paramName]
	if !ok {
		return false, fmt.Errorf("%s: %s.%s not found", INVALID_PARAMETER, paramParentName, paramName)
	}
	parameterValue, ok := parameter.(bool)
	if !ok {
		return false, fmt.Errorf("%s: %s.%s '%v' is not a boolean", INVALID_PARAMETER, paramParentName, paramName, parameter)
	}

	return parameterValue, nil
}

func loadParameterAsString(parameterMap map[interface{}]interface{}, paramParentName string, paramName string, notEmpty bool) (string, error) {
	parameter, ok := parameterMap[paramName]
	if !ok {
		return "", fmt.Errorf("%s: %s.%s not found", INVALID_PARAMETER, paramParentName, paramName)
	}
	parameterValue, ok := parameter.(string)
	if !ok {
		return "", fmt.Errorf("%s: %s.%s '%v' is not a string", INVALID_PARAMETER, paramParentName, paramName, parameter)
	}
	if notEmpty {
		if len(parameterValue) == 0 {
			return "", fmt.Errorf("%s: %s.%s cannot be empty", INVALID_PARAMETER, paramParentName, paramName)
		}
	}

	return parameterValue, nil
}

func loadParameterAsInt(parameterMap map[interface{}]interface{}, paramParentName string, paramName string, hasMin bool, minValue int, hasMax bool, maxValue int) (int, error) {
	parameter, ok := parameterMap[paramName]
	if !ok {
		return 0, fmt.Errorf("%s: %s.%s not found", INVALID_PARAMETER, paramParentName, paramName)
	}
	parameterValue, ok := parameter.(int)
	if !ok {
		return 0, fmt.Errorf("%s: %s.%s '%v' is not an integer", INVALID_PARAMETER, paramParentName, paramName, parameter)
	}
	if hasMin {
		if hasMax {
			if parameterValue < minValue || parameterValue > maxValue {
				return 0, fmt.Errorf("%s: %s.%s  out of range [%d, %d]", INVALID_PARAMETER, paramParentName, paramName, minValue, maxValue)
			}
		} else {
			if parameterValue < minValue {
				return 0, fmt.Errorf("%s: %s.%s  must be >= %d", INVALID_PARAMETER, paramParentName, paramName, minValue)
			}
		}
	} else {
		if hasMax {
			if parameterValue > maxValue {
				return 0, fmt.Errorf("%s: %s.%s  must be < %d", INVALID_PARAMETER, paramParentName, paramName, maxValue)
			}
		}
	}

	return parameterValue, nil
}

func loadParameterAsFloat(parameterMap map[interface{}]interface{}, paramParentName string, paramName string, hasMin bool, minValue float64, hasMax bool, maxValue float64) (float64, error) {
	parameter, ok := parameterMap[paramName]
	if !ok {
		return 0, fmt.Errorf("%s: %s.%s not found", INVALID_PARAMETER, paramParentName, paramName)
	}
	parameterValue, ok := parameter.(float64)
	if !ok {
		return 0, fmt.Errorf("%s: %s.%s '%v' is not a float", INVALID_PARAMETER, paramParentName, paramName, parameter)
	}
	if hasMin {
		if hasMax {
			if parameterValue < minValue || parameterValue > maxValue {
				return 0, fmt.Errorf("%s: %s.%s  out of range [%f, %f]", INVALID_PARAMETER, paramParentName, paramName, minValue, maxValue)
			}
		} else {
			if parameterValue < minValue {
				return 0, fmt.Errorf("%s: %s.%s  must be >= %f", INVALID_PARAMETER, paramParentName, paramName, minValue)
			}
		}
	} else {
		if hasMax {
			if parameterValue > maxValue {
				return 0, fmt.Errorf("%s: %s.%s  must be < %f", INVALID_PARAMETER, paramParentName, paramName, maxValue)
			}
		}
	}

	return parameterValue, nil
}
